unblob/processing.py

```diff
@@ -191,6 +192,22 @@ def remove_inner_chunks(chunks: List[ValidChunk]) -> List[ValidChunk]:
     return outer_chunks
 
 
+def fix_chunk_overlaps(chunks: List[ValidChunk]) -> List[ValidChunk]:
+    """Fix potential overlaps between valid chunks. This can happen with
+    compression streams followed by a chunk starting with data that is
+    considered valid compressed data by the compression algorithm."""
+
+    if not chunks:
+        return []
+
+    sorted_by_offset = sorted(chunks, key=attrgetter("start_offset"))
+    for chunk, next_chunk in pairwise(sorted_by_offset):
+        if chunk.end_offset > next_chunk.start_offset:
+            chunk.end_offset = next_chunk.start_offset - 1
```

Comment:
I will be honest, I do not like this idea.

Once we have calculated and returned with a `ValidChunk`, with valid start and end offset, I think we should **never** modify those values, otherwise the whole operation goes to some weird, unexpected roads.

In my option:  if there are overlapping chunks, it always means some problem.  In those cases we should emit some error messages to the user and possibly remove the erroneous chunk from the valid chunk list (....remove both? ....) and later check the handlers and fix them. 

Of course, you know better the nature of these things than me, but I really don't want to mix things and do these modifications later - this is what the search_chunks / calculate_chunk is for. 

I personally would be really happy if we could solve this in some other way - but of course, no idea how :sweat_smile:

Reply:
I can only agree with you that this is not the best of solutions. This situation can only happen with compression streams so I looked at each compression handler we currently have. The only ones that may lead to such a situation are LZW (compress), and LZMA. With others we can derive the actual offset from header fields, match on a unique end marker (e.g. bzip2), or the stream won't be valid if we read past it (e.g. gzip with isize and crc32). So for those overlapping chunks must never happen.

For LZMA and LZW there is no way we can make these handlers find the actual end offset if the next chunk starts with data that is legit to the decompressor. Overlaps can and will happen. It's just the nature of these compression formats.

We could just get rid of the overlapping chunk but I don't like the idea, and I kinda like the fact that we can find the real end offset of a compressed chunk by using the next valid chunk start address (seriously it's great).

What we could do to limit the domino effect of such a change would be to check whether the handler that created the overlapping chunk is one of our known "bad boys" (unix compress or LZMA). If it's one of those, we re-calculate the offset like I'm doing now. What do you think ?

Reply:
Ok, I understand this is not the perfect way to do it, but if there is no anything better, of course I am happy with your idea. Limiting this for only the "bad bays" is a great addition, I would love that. And also to see some tests for this function.

Thanks for the improvement :+1: